### PR TITLE
Move push to codecov to its own stage

### PR DESCRIPTION
**Pull request by santisoler**
#### Summary

Make each test job in Azure to publish the coverage reports as artifacts. Create a new stage for pushing the coverage reports to Codecov.

#### PR Checklist
* [ ] If this is a work in progress PR, set as a Draft PR
* [ ] Linted my code according to the [style guides](https://docs.simpeg.xyz/content/getting_started/contributing/code-style.html).
* [ ] Added [tests](https://docs.simpeg.xyz/content/getting_started/practices.html#testing) to verify changes to the code.
* [ ] Added necessary documentation to any new functions/classes following the
      expect [style](https://docs.simpeg.xyz/content/getting_started/practices.html#documentation).
* [ ] Marked as ready for review (if this is was a draft PR), and converted
      to a Pull Request
* [ ] Tagged ``@simpeg/simpeg-developers`` when ready for review.

#### Additional information

This allows us to download coverage reports locally, and also prevents codecov for reporting back failures while tests are still being run (and all coverage reports haven't been pushed yet).

### CODE REVIEW

File: .ci/azure/codecov.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    displayName: Upload to Codecov
+    steps:
+      # Checkout simpeg repo. Codecov needs the repo in the file system for
+      # uploading coverage reports.
+      - checkout: self
+        displayName: "Checkout repository"
+
+      - task: DownloadPipelineArtifact@2
+        inputs:
+          patterns: "**/coverage-*.xml"
```
Comment (santisoler) [inline .ci/azure/codecov.yml:14]:
```suggestion
          patterns: "coverage-*/coverage-*.xml"
```

File: .ci/azure/codecov.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        displayName: "Download coverage artifacts"
+
+      - bash: ls -la $(Pipeline.Workspace)/**/coverage-*.xml
```
Comment (santisoler) [inline .ci/azure/codecov.yml:17]:
```suggestion
      - bash: ls -la $(Pipeline.Workspace)/coverage-*/coverage-*.xml
```

**Comment by jcapriot** [comment]
Is your goal to avoid the pre-mature commenting issue you had on #1483? Codecov is supposed to wait until all other GitHub checks are finished before it does anything, so it might’ve been an issue with the GitHub checks, (or a cancelled test potentially).

**Comment by santisoler** [comment]
Codecov does report back after the first test job finishes, this is why we get red crosses in the PR status while tests are still running. One of my goals is to avoid that. Also, to separate the tests from the coverage push: if the upload to codecov fails we need to rerun the failed jobs, and therefore rerun the corresponding tests. By moving the push to another stage, we can just rerun it if it fails without the need to rerun tests.

**Pull request merged**